Add user authentication

mbramson opened the pull request:
Add user authentication.

Add the session and registration paths.

Add ex-machina for testing and comeonin for password hashing.

Any accounts that currently exist before the migration in this PR will be assigned a password of "password" (which they can change by editing their user).

Closes #5

mbramson committed:
5: Add user authentication

Add plug for user authentication

Users can now register and login.

Logged in users can create other users and update their own password.

Migration adds a default password of "password" to each existing user.

mbramson commented:
So I think that I'm going to lead Authentication tokens (like https://github.com/ueberauth/guardian) to a separate PR. It's a great exercise, and this PR is already fairly massive as it is.

daveshah commented on web/models/user.ex:
```diff
@@ -6,19 +6,47 @@ defmodule Pairmotron.User do
+    |> validate_length(:password_confirmation, min: @minimum_password_length)
+    |> validate_confirmation(:password)
+    |> generate_password
+  end
```
Could these two (`changeset` and `registration_changeset` )use a common private function to reduce the duplication?

mbramson replied:
I was thinking about that actually. The answer is yes.

mbramson replied:
This has been done. Let me know what you think of the implementation.

daveshah commented on web/router.ex:
```diff
@@ -9,14 +9,29 @@ defmodule Pairmotron.Router do
+
+    get "/registration", RegistrationController, :new
+    post "/registration", RegistrationController, :create
+
```
could we replace this with `resources "registration", RegistrationController, only: [:new, :create]` ?

mbramson replied:
Definitely.

daveshah commented on web/models/user.ex:
```diff
@@ -6,19 +6,47 @@ defmodule Pairmotron.User do
+    |> generate_password
+  end
+
+  def generate_password(changeset) do
```
should this be private?

mbramson replied:
Yep. Before, it was actually called in the user and registration create and new routes, which necessitated it being public. Then realized it should probably just be part of the changeset itself.

daveshah commented on web/models/user.ex:
```diff
@@ -6,19 +6,47 @@ defmodule Pairmotron.User do
+  def generate_password(changeset) do
+    case changeset do
+      %Ecto.Changeset{valid?: true, changes: %{password: pass}} ->
+        put_change(changeset, :password_hash, Comeonin.Bcrypt.hashpwsalt(pass))
```
It feels weird that we're sprinkling this dependency around (also [here](https://github.com/ElixirCLE/pairmotron/pull/31/files#diff-e8a7c764b6ea7aeebe12d2ecd2313a5aR36) ) I'm wondering if Password type stuff should be extracted into its own module - putting an interface in front of this dependency `¯\_(ツ)_/¯`

mbramson replied:
That's a good call-out. My thought process in not making it its own module was admittedly partly due to the various projects/tutorials I looked at implementing this in the user model. It was also placed here partly because it modifies a changeset (and only a User changeset). Another (not as great) argument for it being here is that it really isn't a ton of logic.

mbramson replied:
So now that there is a common_changeset method that only calls generate_password once I think the argument for keeping here is slightly stronger

mbramson commented:
@daveshah I still need to adapt the tests that test generate_password directly to just test the changesets themselves, but otherwise I think I've addressed your concerns.

edit: Looking at the tests that I have so far, I think the generate_password functionality is covered by the SessionController tests.

mbramson commented:
Disabled the ability for users to set new passwords for users that are not themselves per @ericworkman 's suggestions. I'm leaving the password and password_confirmation fields on the user edit form universally for now, they just cannot be used when looking at a different user (an error is thrown).

mbramson commented:
Updated the database migration so that it assigns a password of "password" to all users that exist before the migration.

ericworkman commented on web/controllers/user_controller.ex:
```diff
@@ -62,4 +72,11 @@ defmodule Pairmotron.UserController do
     |> redirect(to: user_path(conn, :index))
   end
+
+  defp current_assigned_user_id(conn) do
```
Near duplicate of https://github.com/ElixirCLE/pairmotron/pull/31/files#diff-c9ec1fc6e91a7bd967c8f8506027d1f9R27.

See the note for https://github.com/ElixirCLE/pairmotron/pull/31/files#diff-7dc24a26994057cb34cfbdb463c15b3fR44

ericworkman commented on web/controllers/user_controller.ex:
```diff
@@ -41,6 +41,16 @@ defmodule Pairmotron.UserController do
     user = Repo.get!(User, id)
     changeset = User.changeset(user, user_params)
 
+    if current_assigned_user_id(conn) != id && changeset.changes[:password] do
```
Since user is from `Repo.get!`, it should be safe to use `user.id` instead of `id` here. Then you don't have the `to_string` cast in `current_assigned_user_id`.

mbramson replied:
I am going to leave this for a separate PR, but you are correct. This should be refactored.

ericworkman commented on web/templates/layout/app.html.eex:
```diff
@@ -20,6 +20,14 @@
+              <%= case @conn.assigns[:current_user] do %>
+                <% nil -> %>
+                  <%= link "Login", class: "btn btn-success", role: "button", to: session_path(@conn, :new) %>
+                <% current_user -> %>
```
Isn't used, so it should be `_current_user`

mbramson replied:
Great call.

ericworkman reviewed (approved):
Looks good. Fix these minor things in this PR or in others.
